basic compile support for grouped_mm

bdhirsh opened the pull request:
grouped_mm is used in torchtitan, this adds just enough support in compile to allow inductor to lower it as a fallback kernel. I imagine that at some point in the future it may be valuable to get inductor to support templating grouped_mm, although this PR just provides basic support. cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @ngimel @eellison 



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153384



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

bdhirsh committed:
basic compile support for grouped_mm

[ghstack-poisoned]

bdhirsh committed:
Update on "basic compile support for grouped_mm"

grouped_mm is used in torchtitan, this adds just enough support in compile to allow inductor to lower it as a fallback kernel. I imagine that at some point in the future it may be valuable to get inductor to support templating grouped_mm, although this PR just provides basic support. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov ngimel eellison 






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "basic compile support for grouped_mm"

grouped_mm is used in torchtitan, this adds just enough support in compile to allow inductor to lower it as a fallback kernel. I imagine that at some point in the future it may be valuable to get inductor to support templating grouped_mm, although this PR just provides basic support. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov ngimel eellison 






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented on test/inductor/test_torchinductor.py:
```diff
@@ -13741,6 +13742,50 @@ def forward(
 
+        @skipCUDAIf(not SM90OrLater, "Requires sm90")
+        @requires_gpu()
+        @config.patch(implicit_fallbacks=True)
```
I removed the lowering I added in inductor - I realized that implicit_fallbacks actually defaults to `True`, but was set to false in this test suite

bdhirsh replied:
good point. maybe @ngimel would know? (does grouped_mm only support contiguous inputs?)

ngimel replied:
For grouped_mm we don't other than one of the dimensions should be contiguous, similar to regular mm.

eellison replied:
@bdhirsh can you add `make_fallback(aten._grouped_mm, require_dense)`

ngimel replied:
@eellison what would be the correct way of handling this for custom ops provided by third-party libraries?

eellison replied:
For custom ops provided by third-party libraries, we are trying to match the eager strides by default. @zou3519 is going to turn that on (i forget if he did or not yet).

Otherwise we have these tags: https://github.com/pytorch/pytorch/blob/236b08cbf83edcd168aaa384e2603dbf8d2acc2f/aten/src/ATen/native/tags.yaml#L45-L67.

ngimel replied:
Ah interesting, so there's no "require_dense" equivalent?

eellison replied:
Theres not, we could add one if someone wanted

zou3519 replied:
> For custom ops provided by third-party libraries, we are trying to match the eager strides by default. @zou3519 is going to turn that on (i forget if he did or not yet).

This is on by default already

etaf commented on test/inductor/test_torchinductor.py:
```diff
@@ -13741,6 +13742,50 @@ def forward(
+                    a, b.transpose(-2, -1), offs=offs, out_dtype=out_dtype
+                )
+
+            device = "cuda"
```
Hi, may I suggest we replace the hardcode '"cuda"' with `GPU_TYPE`? This test case is decorated with `requires_gpu` and will also runs on Intel GPU, which will fail with cuda not compiled error.
You can also use `requires_cuda` here.

bdhirsh replied:
I'm going to come back to this PR in a few days but good call out, will update

bdhirsh committed:
Update on "basic compile support for grouped_mm"

grouped_mm is used in torchtitan, this adds just enough support in compile to allow inductor to lower it as a fallback kernel. I imagine that at some point in the future it may be valuable to get inductor to support templating grouped_mm, although this PR just provides basic support. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov ngimel eellison 






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "basic compile support for grouped_mm"

grouped_mm is used in torchtitan, this adds just enough support in compile to allow inductor to lower it as a fallback kernel. I imagine that at some point in the future it may be valuable to get inductor to support templating grouped_mm, although this PR just provides basic support. cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov ngimel eellison 






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -r

bdhirsh committed:
Update

[ghstack-poisoned]
